Add development.md (#4712)
onflow/flow-go: DEVELOPMENT.md

> # Developement

Reviewer: ```suggestion
# Development
```

# Developement

## Feature Release Flow

### Branching Strategy

- The `master` branch is intended to only contain features for the immediately upcoming release, whether it is a Height Coordinated Upgrade or a Spork
- During development, all features should live on a feature branch.

> - For small features, this will be simple working branch, i.e. `kan/1-fix-known-issue`

Reviewer: ```suggestion
- For small features, this will be simple working branch. These branches have the naming scheme `<contributor>/<issue #>-<short description>, for example `kan/123-fix-known-issue`
```

> - For larger features, this may be a shared feature branch with other team members.

Reviewer 2: Maybe mention the feature branch naming scheme here 
```suggestion
- For larger features, this may be a shared feature branch with other team members. Feature branches have the naming scheme `feature/<name>`.
```

### Upgrade Path Eligibility

> - When a feature branch is ready to be merged, it must be determined if the desired upgrade path onto Mainnet. The options are:

Reviewer 2: The second half reads like the order is off or something is missing. Maybe
```suggestion
- When a feature branch is ready to be merged, the desired upgrade path onto Mainnet must be determined (if any). The options are:
```

    - Height Coordinated Upgrade (HCU)
        - No protocol level breaking changes
        - No state migrations

>         - Execution state/path changes allowed

Reviewer 2: What does this mean in particular. Changes to how execution state is stored need a state migration / spork. 
Maybe clarify to avoid potential misunderstanding and problems due to it.

        - Resource optimizations are okay
        - Cadence upgrades which could cause execution state fork (likely any Cadence upgrade except for trivial changes)
    - Spork
        - Protocol level breaking change
        - State migrations required
- All HCU upgrades can go directly into the `master` branch

> - All spork upgrades must live on a their own feature branch until the last HCU before the spork as been performed (usually approximately 1 month before the Spork.

Reviewer: ```suggestion
- All spork upgrades must live on their own feature branch until the last HCU before the spork as been performed (usually approximately 1 month before the Spork.
```

    - It is the responsibility of the DRI to keep this feature branch in a mergeable state

>     - Suggestion: once a sprint, merge `master` into the feature branch

Reviewer 2: ```suggestion
    - Suggestion: once a sprint, merge `master` into the feature branch. More frequent merges are easier, as they avoid complex conflict resolutions
```

### End of Release Cycle

> - At the end of every release cycle, we will tag a commit that includes all desired features to be Released

Reviewer 2: ```suggestion
- At the end of every release cycle, we will tag a commit that includes all desired features to be released
```

> - This commit will be tagged according to semantic versioning guidelines

Reviewer 2: Maybe link. @janezpodhostnik had a doc

- Release notes must be written up, describing all features included in this tag

### Benchmark Testing

[Benchmarking](https://www.notion.so/Benchmarking-e3d89e3aadb44b0787da9bb7703b0dae?pvs=21)

- All features on the release candidate tag must undergo testing on `benchmarknet`

### Testnet

> - The current schedule is the Wednesdays two weeks before the corresponding mainnet spork

Reviewer 2: ```suggestion
- The current schedule is the Wednesdays two weeks before the corresponding Mainnet spork
```

> - The current schedule is the Wednesdays two weeks before the corresponding mainnet spork

Reviewer 3: ```suggestion
- The current schedule is the Wednesday two weeks before the corresponding mainnet spork
```

> - The current schedule is the Wednesdays two weeks before the corresponding mainnet spork

Reviewer 3: ```suggestion
- The current schedule is the Wednesday two weeks before the corresponding Mainnet spork
```

- Features should aim to live on Testnet for at least a week before making it to Mainnet

### Mainnet

> - Features must live on Testnet for two week before making it to Mainnet

Reviewer 3: This conflicts with the previous statement:
> - Features should aim to live on Testnet for at least a week before making it to Mainnet

should this be updated to one week, or does it need more qualification?

> - Features must live on Testnet for two week before making it to Mainnet

Reviewer 3: This conflicts with the previous statement:
> - Features should aim to live on Testnet for at least a week before making it to Mainnet

should this be updated to one week, or does it need more qualification?

Author: updated both to 2 weeks

> - The current schedule is the Wednesday two week after the Testnet Spork

Reviewer 3: ```suggestion
- The current schedule is the Wednesday two weeks after the Testnet Spork
```

## Breaking Change Classifications

### Acceptable Changes for HCU

- All backward compatible changes
- Breaking changes only pertaining to the execution of future transactions
    - Many Cadence related breaking changes would fall in this category
    - FVM changes may also fall here
- Breaking changes only pertaining to the verification of future transactions

### Spork only changes

- Any change that requires a state migration
    - i.e. something changing in how the historical state will be read and interacted with
- Any change that would break the communication contract between nodes
    - e.g. Addition of a new REQUIRED field in a message structure
    - Removal of a REQUIRED field in a message structure
    - Removal of a REQUIRED channel in libp2p
    

>     Generally, *all* the fields in our node-to-node messages are required. For BFT reasons we avoid optional fields, as they add surface for spamming attacks, impose additional consistency requirements, and they often add security vulnerabilities w.r.t. the message hash/ID. I would strongly discourage working temporarily with optional fields, as this adds technical debt of needing to change those fields later to being required. 

Reviewer 2: Add line breaks
